src/v/iceberg/merge_append_action.cc

```diff
@@ -0,0 +1,552 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+#include "iceberg/merge_append_action.h"
+
+#include "base/units.h"
+#include "base/vlog.h"
+#include "iceberg/logger.h"
+#include "iceberg/manifest.h"
+#include "iceberg/manifest_file_packer.h"
+#include "iceberg/manifest_list.h"
+#include "iceberg/snapshot.h"
+#include "iceberg/table_requirement.h"
+#include "iceberg/values_bytes.h"
+#include "random/generators.h"
+
+#include <iterator>
+#include <limits>
+
+namespace iceberg {
+
+namespace {
+manifest_path get_manifest_path(
+  const ss::sstring& location, const uuid_t& commit_uuid, size_t num) {
+    return manifest_path{
+      fmt::format("{}/metadata/{}-m{}.avro", location, commit_uuid, num)};
+}
+manifest_list_path get_manifest_list_path(
+  const ss::sstring& location,
+  snapshot_id snap_id,
+  const uuid_t& commit_uuid,
+  size_t num) {
+    return manifest_path{fmt::format(
+      "{}/metadata/snap-{}-{}-{}.avro", location, snap_id(), commit_uuid, num)};
+}
+
+action::errc to_action_errc(manifest_io::errc e) {
+    switch (e) {
+    case manifest_io::errc::failed:
+        return action::errc::io_failed;
+    case manifest_io::errc::shutting_down:
+        return action::errc::shutting_down;
+    case manifest_io::errc::timedout:
+        // NOTE: treat IO timeouts the same as other IO failures.
+        // TODO: build out retry logic.
+        return action::errc::io_failed;
+    }
+}
+
+snapshot_id random_snap_id() {
+    return snapshot_id{random_generators::get_int<int64_t>(
+      0, std::numeric_limits<int64_t>::max())};
+}
+
+snapshot_id generate_unused_snap_id(const table_metadata& m) {
+    auto sid = random_snap_id();
+    if (!m.snapshots.has_value() || m.snapshots->empty()) {
+        return sid;
+    }
+    // Repeatedly try to generate a new snapshot id that isn't used already.
+    const auto& snaps = *m.snapshots;
+    while (std::find_if(
+             snaps.begin(),
+             snaps.end(),
+             [sid](const snapshot& snap) { return snap.id == sid; })
+           != snaps.end()) {
+        sid = random_snap_id();
```

Comment:
how many snapshots do you expect here?

Reply:
This _should_ be configurable, but the default behavior I believe is that we should clean up snapshots after 5 days. I think we have the liberty to remove snapshots every hour if we want to (so maybe on the order of dozens or hundreds?), IIUC.

https://www.tabular.io/apache-iceberg-cookbook/data-operations-snapshot-expiration/ has some info on snapshot expiration. This will need to be done for applying table retention.

Reply:
got it. i think this should be fine. things that i was thinking are cryptographic random vs normal random generator. was also thinking like oh if we get a bunch of conflicts should we just choose the next max value. i don't think it matters at all, tho. 64 bit is pretty big space.